api-report/tree.api.md

```diff
@@ -1094,9 +1094,14 @@ function rebase<TNodeChange>(change: Changeset<TNodeChange>, base: TaggedChange<
 // @public
 export function recordDependency(dependent: ObservingDependent | undefined, dependee: Dependee): void;
 
+// @public
+export type RecursiveReadonly<T> = T extends undefined | null | boolean | string | number | Function | Symbol ? T : T extends Opaque<Brand<any, string>> ? T : T extends Map<infer K, infer V> ? ReadonlyMap<RecursiveReadonly<K>, RecursiveReadonly<V>> : T extends Set<infer K> ? ReadonlySet<RecursiveReadonly<K>> : T extends (infer V)[] ? readonly RecursiveReadonly<V>[] : {
```

Comment:
This is kinda ugly and I fear will make devx for auto-complete, auto-implement, hover and error messages suck when we use it.

I also think we want to limit it a bit more. This shouldn't be allowed on functions, and probably should not be used on types with methods except for the types it explicitly handles (set map and array) since arbitrary methods might mutate.

simple example: using the implement interface helper you get:

```typescript
interface Foo {
    x: Foo
}

class FooX implements RecursiveReadonly<Foo>{
    x: { readonly x: ...; };
}
```

If fails to figure out the type and puts in "...". Mousing over `type T = RecursiveReadonly<Foo>;` lists the type as `type T = {
    readonly x: ...;
}`

Since we expect external consumers to work with readonly deltas, and not mutable deltas, I think the approach of making the readonly one have these issues is not great.

I have generally taken the approach of making the public types all readonly, and having the code that builds them use local mutable (often inferred) types instead. That approach might not work great here.

Also its not clear if the contract is that you can't change readonly deltas, or that they are constant (no one will change them). If they really are immutable, it might be nice to have an explicit type for that.

Another option is to make deltas opaque in these APIs, and force you to use the visitor, which also happens to remove the mutability.

Reply:
Example of bad error from this approach:
```
(property) TaggedChange<TChangeset>.change: TChangeset
Type 'TChangeset' is not assignable to type 'RecursiveReadonly<TChangeset>'.
  Type 'TChangeset' is not assignable to type 'TChangeset extends string | number | boolean | Function | TChangeset | Symbol ? TChangeset : TChangeset extends BrandedType<...> ? TChangeset : TChangeset extends Map<...> ? ReadonlyMap<...> : TChangeset extends Set<...> ? ReadonlySet<...> : TChangeset extends (infer V)[] ? readonly RecursiveReadonly<...>[] : { read...'.ts(2322)
lib.es5.d.ts(1441, 24): The expected type comes from the return type of this signature.
```

I'd like to avoid these kinds of things. Manually declaring the original types as immutable and occasionally making (and not package exporting) partially mutable (but usually not deeply mutable) types for internal use as needed seems like it might be a lot cleaner. Its a little redundent, but should be safe from the two version diverging if we make add a type assert they are assignable as required, and/or make one extend the other.

Reply:
The current contract is you can't change them, but the implementation does not mutate them, so they are effectively constant.
I can't imagine us ever needing to mutate them, so I'd be okay with reflecting the constness in the contract. The thing that's not clear to me is how that is different from `RecursiveReadonly`. Is it just a different name (e.g. `RecursiveImmutable`) that we document to be immutable as opposed to solely readonly?

I'll try going the "start with the readonly types" route and see what that's like.